# Learn Python Basics

Well, hello everybody and welcome to our workshop, the purpose of which is to expand and improve the capacity of our robot. I hope you guys are ready because this is so much fun! Throughout this workshop, you will learn so many new and exciting things so don’t forget to take notes and don’t be shy, the supervising staff is here to answer all your questions. Are you ready? Let’s go!  

## What is Python exactly ?
Python is an interpreted high-level programming language that's focused on its readability and its simple synthax, which facilitates a lot collaborations.\
It's widly used in the research domains and the world of mathematics, mostly by Data Scientists and Researcher in a lot of domains.

## Variables

Houston, we have a problem. Indeed, we sent to our robot a file. Unfortunately, the file is full of incorrect syntaxes, so we need you!  

For your first mission, you must correct errors in the syntax and fix the algorithm if it’s not working. 
[Learn Python Variable](https://www.w3schools.com/python/python_variables.asp)

In [1]:
max = 10
goal = 15
victory = "Success !"
retry = "Not yet"

for i in range(0, max)
    if (i == goal) {
        print(i, victoy)
        break
    }
    print(i, retry)

SyntaxError: invalid syntax (346966306.py, line 6)

#### You must find "Success !" when the code working !

## Operations

Alright! Great job!
Now, we want our robot to execute a bunch of operation and the first thing we need is our robot to express itself. This is where you come in.

We need you to allow the robot to say its first word by concatenating two strings! 

Create a function `concatenation_string` for assembly two string just bellow (first and second). 

[Learn Python Operators](https://www.w3schools.com/python/python_operators.asp)
[Standard operators Python](https://docs.python.org/3/library/operator.html)

In [2]:
first = "Hello"
second = "World !"

## Your function

##Print result here

4


#### If you get "Hello World!" you've passed this step!

Ok, so you are clearly a professional! Let’s see if we can make things less easy for you because, obviously, you are good for this.  

## Conditions

It’s time for you to create a function that will allow our robot to know how much there is multiple of 3 and 5 between 0 and 7854. For this kind of function, the modulo sign “%” is important and necessary so don’t be afraid to find the signification of it.  

[Learn Python Conditions](https://www.w3schools.com/python/python_conditions.asp)
[Standard Python Condition](https://python.doctor/page-apprendre-conditions-structures-conditionnelles-if-else-python-cours-debutant)


In [3]:
## Write your function here

## Print your result

3665


Is your result equal to 3665?
Go quickly to the next step!

## List
It is very important to understand how a list works. There are different ways to browse a list.

Here are operations that we can apply on lists or numpy arrays.

I invite you to remember the slicing with the operator:
It's up to you to test and understand what the examples below do.

[Learn Python List](https://docs.python.org/fr/3/tutorial/datastructures.html)

In [4]:
x = [1, 2, 3, 4, 5, 6]
print(len(x))

6


In [5]:
x[:3]

[1, 2, 3]

In [6]:
x[3:]

[4, 5, 6]

In [7]:
x[-2:]

[5, 6]

In [8]:
x.extend([7,8])
x

[1, 2, 3, 4, 5, 6, 7, 8]

In [9]:
x.append(9)
x

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [10]:
y = [10, 11, 12]
listOfLists = [x, y]
listOfLists

[[1, 2, 3, 4, 5, 6, 7, 8, 9], [10, 11, 12]]

In [11]:
y[1]

11

In [12]:
z = [3, 2, 1]
z.sort()
z

[1, 2, 3]

In [13]:
z.sort(reverse=True)
z

[3, 2, 1]

**Now is the time to test the robot's accuracy and comprehesion.**

Writes a program that allows it to delete all words with a size other than seven characters.

In [14]:
my_list = [
    "It's   ", "Hello", "Yet", "Maybe"
, "No", "For", "Yes", "Success", "Oups", "Above", "AIIIe", "AHAH"]

# your code goes here

# Print result

Yes, it's still a success.
We continue towards our next step, the modularity!

## Modularity
In this language you can call differents librairies or function in differents files.

Write two function and use the module "math":

Example:

value = math.sum(3 + 4)
print(value) ## Commentaire: 7

You must create function that return the square root of parameter number and another that return the exponential.

[Learn Module Python](https://docs.python.org/3/reference/import.html)
[Standard Module Python](https://docs.python.org/3/tutorial/modules.html)

In [6]:
## EXAMPLE CODE

##Import your module (you can precise the function that you want)
import math

nbr_square = 65
base = 4
exponent = 18

## Set variable and call the function of your librairies.
## The function fsum return an accurate floating point sum of values in the iterable.
value = math.fsum([.1, .1, .1, .1, .1, .1, .1, .1, .1, .1])

## Print your variable
print(value)

1.0


If you get 8.06225774829855 for the root square of 65 and 68719476736 for the exponential of your base. You have succeeded!

## Special list
After seeing how lists can be manipulated, here we have some other type of list.
We have **Tuple** (Non mutable list) and **Dictionnary** (Keys=>Value list).

We will see only the dictionaries but remember that the tuples exist.

### Dictionnairies
Dictionnaries are some kind of Keys => Values list, as you can have in JSON configuration files.
[Learn Python Dictionnaries](https://docs.python.org/3/tutorial/datastructures.html)

In [18]:
dictionnary = {"robot": "This one is a little robot", "big_robot": "This one is a BIG robot", "advanced_robot": "This one is a smart robot"}
print(dictionnary["robot"])

selector = ["big_robot", "advanced_robot"]
for i in selector:
    print(dictionnary[i])

operators = {"+" : lambda x,y:x+y, "-" : lambda x,y:x-y}
calcul = "1+3"
operators[calcul[1]](int(calcul[0]), int(calcul[2]))

This one is a little robot
This one is a BIG robot
This one is a smart robot


4

## Classes and Heritages

Classes is what we call **Oriented Object Programming** (Or OOP) and is essential in a vast number of languages.
To vulgarize it, classes are sort of mold used to create **object**. Once you've the molds, you can create as many objects of the same type as you want. This is used in every import you do, any functions from libraries are **methods** from classes.
There names are often written with a uppercase at the beginning. 

If you want to get deeper in this notion, I highly recommend you to search on Internet. It's a highly well explained subject that's super well explained.
[Python classes doc](https://docs.python.org/3/tutorial/classes.html)\
[OpenClassroom course](https://openclassrooms.com/fr/courses/7150616-apprenez-la-programmation-orientee-objet-avec-python/7195794-creez-et-utilisez-des-objets-python)


Because an example is more meaningful than words, here's an exemple.
A class made for a robot will have its pure concept of a robot as the mold and your unique handmade robot as an object.
We'll have something in these lines.


In [19]:
class Robot:
    # Variable of a class are called attributes.
    # These are accessible from the body of the class with no instanciation at any moment, and can be considered as "global variable".
    cur_robot = []

    # Here's what we call a "constructor", it's the part that create the object itself, using instance atribute.
    # This is also called the "instanciation".
    def __init__(self, name = "unnamed") -> None:    # "name" parameter has a default value "unnamed"
        if (name in self.cur_robot):
            raise ValueError("This robot already exist.")    # Error managing line. If we have two robots with the same name, this will stop the program
        
        self.cur_robot.append(name) # Self is the variable of the instance. It defines the object.
        self.name = name
        self.inventory = []
        
    # "Method" are the functions of an objects. It will use its attributes.
    # It will interact, and changes the instances of the class.
    def add_item(self, item: str):
        if (item not in self.inventory):
            self.inventory.append(item)
    
    def speak(self):
        print("Hello, my name is " + self.name)
   

print(Robot.cur_robot)
Steve = Robot("Steve")
print(Robot.cur_robot)
Steve.speak()

[]
['Steve']
Hello, my name is Steve


## Application
Your turn now, using what you've learn, try to apply and create a class that generates more advanced robots !
These robots will be able to **walk**, manage their **positions** (and so **coordinates** that can be consultated) and most importantly they will be able to talk to other robots !
Good luck and get to us whenever you feel like you didn't understand something.

In [20]:
class AdvancedRobot:
    def __init__(self) -> None:
        pass

    def walk(self) -> None:
        pass

    def give_positions(self) -> list: # This one can print too, its your choice !
        return

    def interact(self, AdvancedRobot) -> None:
        pass

Secondly, we will have classes inheritance. Classes inheritance is simply a child of the first class that will interact with parents specificities, with all its caracteristic and mostly everything it does. It can be customized, and it's really useful when it comes into complex classes.
It's anything harder, there's a lot of specifities but you'll learn it in the long way !

In [21]:
class BigRobot(Robot):
    def __init__(self, name):
        super().__init__(name=name)
        
    def speak(self):
        print("HELLO I'M THE BIG ROBOT, MY NAME IS " + self.name.capitalize())

big_steve = BigRobot("Josh")
big_steve.speak()

HELLO I'M THE BIG ROBOT, MY NAME IS Josh


## Final exercise

Here we are chief ! You come from a long way, but here's the ultimate challenge to apply everything you've learnt. Don't hesitate to recheck anything if you're not sure, and maybe get deeper in a specific notion (And maybe do some Python after this workshop !).

The goal of this final exercise is to manage a factory of robot, that are made for listing and sorting elements. These aren't entirely complete, but they can sort a lot of material, and you've to create and adapt the system to make it extensible to all type of data. Good luck !
Par contre j'ai pas ton temps, donc débrouille toi là. T'es grand, casse pas les couilles.

In [22]:
class Products:
    product_type = None
    products = []

    def __init__(self, products: list, product_type: str) -> None:
        self.product_type = product_type
        self.products = products


class ListingRobot:
    
    def __init__(self) -> None:
        pass
    
    def add_item(self, item: str) -> None:
        pass

    def remove_item(self, item_id: int = 0, item_name: str = None) -> None:
        pass

    def find_item(self, item_id: int = 0, item_name: str = None) -> str:
        pass

    def sort_items(self) -> None:
        pass

    def show_item(self) -> None:
        pass
